sys/cddl/contrib/opensolaris/uts/common/dtrace/dtrace.c

```diff
@@ -13245,7 +13245,7 @@ dtrace_dof_copyin(uintptr_t uarg, int *errp)
 	/*
 	 * First, we're going to copyin() the sizeof (dof_hdr_t).
 	 */
-	if (copyin((void *)uarg, &hdr, sizeof (hdr)) != 0) {
+	if (copyin_c((void * __capability) uarg, &hdr, sizeof (hdr)) != 0) {
```

Comment:
Would this work with a non capability userspace?
I think not.
would  checking `SV_CURPROC_FLAG(SV_CHERI)` make it work?
```c
if (SV_CURPROC_FLAG(SV_CHERI))
   copyin_c((void* __capability)....
else
   copyin((void *) uarg...

```

Reply:
`copyin` vs `copyin_c` is about whether it _uses_ a capability to do the copy. What userspace is doing is generally irrelevant, except I believe for mips only `copyin_c` is able to copy capabilities, although that may have already changed. The real issue is whether `hdr` varies between userspace ABIs.

Reply:
This won't work in a hybrid kernel because `uarg` needs to be an `uintcap_t` since you can't count on fabricating a userspace capability from a kernel capability.  That change needs to percolate up to another level where for hybrid processes the integer pointer is converted to a capability with one of the `__USER_CAP*()` macros and for CheriABI processes the capability from userspace is passed along.

Note that neither `copyin` or `copyin_c` preserve tags. You must use `copyincap` if you want to preserve tags (It only has the equivalent of the `_c` variant).